pkg/sync/syncWorker.go

```diff
@@ -51,57 +68,185 @@ func startSyncWorker(
 }
 
 func (s *syncWorker) start() error {
+	// NOTE: subscriptions can be internally de-duplicated
+	// to avoid race conditions, we first set up the listener for new nodes and then all the existing ones
+	s.subscribeToRegistry()
+
 	nodes, err := s.nodeRegistry.GetNodes()
 	if err != nil {
 		return err
 	}
+
 	for _, node := range nodes {
-		if node.NodeID == s.registrant.NodeID() || !node.IsHealthy || !node.IsValidConfig {
-			continue
-		}
-		s.subscribeToNode(node)
+		s.subscribeToNode(node.NodeID)
 	}
+
 	return nil
 }
 
 func (s *syncWorker) close() {
+	s.log.Debug("Closing sync worker")
+	s.cancel()
 	s.wg.Wait()
+	s.log.Debug("Closed sync worker")
 }
 
-func (s *syncWorker) subscribeToNode(node registry.Node) {
+func (s *syncWorker) subscribeToRegistry() {
 	tracing.GoPanicWrap(
 		s.ctx,
 		&s.wg,
-		fmt.Sprintf("node-subscribe-%d", node.NodeID),
+		"node-registry-listener",
 		func(ctx context.Context) {
-			var err error
-			var conn *grpc.ClientConn
-			var stream message_api.ReplicationApi_SubscribeEnvelopesClient
+			newNodesCh, cancelNewNodes := s.nodeRegistry.OnNewNodes()
+			defer cancelNewNodes()
 			for {
-				if err != nil {
-					s.log.Error(fmt.Sprintf("Error: %v, retrying...", err))
-					time.Sleep(1 * time.Second)
+				select {
+				case <-ctx.Done():
+					return
+				case newNodes, ok := <-newNodesCh:
+					if !ok {
+						// data channel closed
+						return
+					}
+					s.log.Info("New nodes received:", zap.Any("nodes", newNodes))
+					for _, node := range newNodes {
+						s.subscribeToNode(node.NodeID)
+					}
 				}
+			}
 
-				conn, err = s.connectToNode(node)
-				if err != nil {
-					continue
-				}
-				stream, err = s.setupStream(node, conn)
-				if err != nil {
-					continue
-				}
-				err = s.listenToStream(stream)
-				if err != nil {
-					continue
+		})
+}
+
+func (s *syncWorker) subscribeToNode(nodeid uint32) {
+	if nodeid == s.registrant.NodeID() {
+		return
+	}
+
+	s.subscriptionsMutex.Lock()
+	defer s.subscriptionsMutex.Unlock()
+
+	if _, exists := s.subscriptions[nodeid]; exists {
+		// we already have a subscription to this node
+		return
+	}
+
+	s.subscriptions[nodeid] = struct{}{}
+
+	tracing.GoPanicWrap(
+		s.ctx,
+		&s.wg,
+		fmt.Sprintf("node-subscribe-%d", nodeid),
+		func(ctx context.Context) {
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					s.subscribeToNodeInner(ctx, nodeid)
 				}
 			}
+		})
+}
+
+func (s *syncWorker) subscribeToNodeInner(ctx context.Context, nodeid uint32) {
+
+	notifierCtx, notifierCancel := context.WithCancel(ctx)
+
+	node, err := s.setupNodeRegistration(notifierCancel, nodeid)
```

Comment:
I'm understanding everything much better now - I wonder if a few small readability changes might help obtuse people like myself, something like the following:

- Pull out a `NodeConfiguration` struct with `node` and `ctx` stored internally - this replaces `node`, `notifierCtx` and `notifierCancel` here
- `setupNodeRegistration()`->`getCurrentNodeConfiguration()` returning a `NodeConfiguration`, and moved out into the `default` clause of the loop in `subscribeToNode()`.
- `subscribeToNodeInner()`->`subscribeToNodeConfiguration()` that accepts a `NodeConfiguration` argument
- Flatten `handleUnhealthyNode()` into `subscribeToNodeConfiguration()` (would just be `<-nodeConfig.ctx.Done()`)
- Set up the stream in `handleNodeConnection()` using `nodeConfig.ctx`

Will leave it up to your judgment!

Reply:
thank you! It's much cleaner that way!